### PR TITLE
Drawing Classifier: Added support for validation set + fixed training metric reporting

### LINKED ISSUE #1464: Fix Drawing Classifier Training Metric Reporting
- [x] Get rid of smoothed loss in Drawing Classifier.
- [x] Yield statistics at the granularity of a training step.
- [x] Output metrics like accuracy.

### LINKED ISSUE #1493: Add automatic validation set to the Drawing Classifier
To prevent overfitting, we should implement a validation split as part of training in the Drawing Classifier.

### PR DESCRIPTION
**Pull request by shantanuchhabra**
Fixes #1493 
Fixes #1464

### CODE REVIEW

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 1]
```diff
@@ -132,7 +146,7 @@ def create(input_dataset, target, feature=None,
     dataset = _extensions._drawing_classifier_prepare_data(
         input_dataset, feature) if is_stroke_input else input_dataset
 
-    column_names = ['Iteration', 'Loss', 'Elapsed Time']
+    column_names = ['Training Step', 'Loss', 'Validation Accuracy', 'Elapsed Time']
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: I think `Iteration` makes more sense than `Training Step`. Displaying `Training Accuracy` would be good too. I'm not sure there is much value in showing `Loss`.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 2]
```diff
@@ -144,7 +158,27 @@ def create(input_dataset, target, feature=None,
...
+            if verbose:
+                print_validation_track_notification()
+            dataset, validation_dataset = dataset.random_split(
+                TRAIN_VALIDATION_SPLIT, seed=seed)
+        else:
+            validation_dataset = _tc.SFrame()
+    elif validation_set is None:
+        validation_dataset = _tc.SFrame()
+    else:
+        raise TypeError("Unrecognized value for 'validation_set'.")
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: `value` -> `type`.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 3]
```diff
@@ -144,7 +158,27 @@ def create(input_dataset, target, feature=None,
     classes = sorted(classes)
     class_to_index = {name: index for index, name in enumerate(classes)}
 
-    def update_progress(cur_loss, iteration):
+    if isinstance(validation_set, _tc.SFrame):
+        _raise_error_if_not_drawing_classifier_input_sframe(
+            validation_set, feature, target)
+        is_validation_stroke_input = (validation_set[feature].dtype != _tc.Image)
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: It doesn't look like `is_validation_stroke_input` is getting used.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 4]
```diff
@@ -42,9 +43,13 @@ def _raise_error_if_not_drawing_classifier_input_sframe(
     if len(dataset) == 0:
         raise _ToolkitError("Input Dataset is empty!")
 
-def create(input_dataset, target, feature=None, 
+def print_validation_track_notification():
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: I'm not really understanding why this is its own function. It should probably be a hidden function.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 5]
```diff
@@ -196,26 +239,42 @@ def update_progress(cur_loss, iteration):
...
+        for validation_batch in validation_loader:
+            validation_batch_data = _mx.gluon.utils.split_and_load(
+                validation_batch.data[0], ctx_list=ctx, batch_axis=0)
+            validation_batch_label = _mx.gluon.utils.split_and_load(
+                validation_batch.label[0], ctx_list=ctx, batch_axis=0)
+            outputs = []
+            for x, y in zip(validation_batch_data, validation_batch_label):
+                z = model(x)
+                outputs.append(z)
+            # Updates internal evaluation
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Not really understanding this comment

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 1 above)
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: I avoided saying "Iteration" because of the `max_iterations` parameter where an iteration is used to refer to an epoch. What do you think? I'll add a Training Accuracy column.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 3 above)
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Thanks for the catch, the next line should use it.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 2 above)
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Done.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 5 above)
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Updated.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 1 above)
Comment (TobyRoseman) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Oh, I'm confused. I thought you were printing out one row per iteration/epoch. . . . .  I would recommend doing that rather than print out every ten seconds or so.

**Comment by TobyRoseman** [comment]
Take a look a the [ProgressTablePrinter util](https://github.com/apple/turicreate/blob/master/src/unity/python/turicreate/util/progress_table_printer.py) I just checked. I'm using it [in the sound classifier](https://github.com/apple/turicreate/blob/23e1e8be55f18580939df768c6561c2e22e48ad8/src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py#L108) and it makes things easier and cleaner. Perhaps you want to use it here too.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 1 above)
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Done.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 6]
```diff
@@ -42,9 +43,9 @@ def _raise_error_if_not_drawing_classifier_input_sframe(
     if len(dataset) == 0:
         raise _ToolkitError("Input Dataset is empty!")
 
-def create(input_dataset, target, feature=None, 
+def create(input_dataset, target, feature=None, validation_set='auto',
             pretrained_model_url=None, batch_size=256, 
-            num_epochs=100, max_iterations=0, verbose=True):
+            num_epochs=100, max_iterations=0, seed=None, verbose=True):
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: I'm not sure there is a lot of value to having a `seed` parameter if it's just being used for the train/validation set split. If the user wants reproducibility they can construct the validation set themselves before calling create. If you do want a `seed` parameter, it should be mentioned in the docstring.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 7]
```diff
@@ -132,7 +142,7 @@ def create(input_dataset, target, feature=None,
     dataset = _extensions._drawing_classifier_prepare_data(
         input_dataset, feature) if is_stroke_input else input_dataset
 
-    column_names = ['Iteration', 'Loss', 'Elapsed Time']
+    column_names = ['Training Step', 'Loss', 'Train Accuracy', 'Validation Accuracy', 'Elapsed Time']
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: So what happens if the user calls `create` with `validation_set=None`? Is that column just empty? It would probably be better for that column to just not appear.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 6 above)
Comment (nickjong) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Let's just be consistent with our other toolkits for now

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 8]
```diff
@@ -196,39 +209,63 @@ def update_progress(cur_loss, iteration):
...
     state = {
         '_model': model,
         '_class_to_index': class_to_index,
         'num_classes': len(classes),
         'classes': classes,
         'input_image_shape': (1, BITMAP_WIDTH, BITMAP_HEIGHT),
         'batch_size': batch_size,
-        'training_loss': cur_loss,
-        'training_time': training_time,
+        'training_loss': train_loss,
```
Comment (nickjong) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Shouldn't we also include the training accuracy, and validation accuracy if available?

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 9]
```diff
@@ -196,39 +209,63 @@ def update_progress(cur_loss, iteration):
...
+                # Computes softmax cross entropy loss.
+                loss = softmax_cross_entropy(z, y)
+                # Backpropagate the error for one iteration.
+                loss.backward()
+                train_outputs.append(z)
+            # Compute training accuracy
+            train_accuracy.update(train_batch_label, train_outputs)
+
+        # Compute validation accuracy
+        for validation_batch in validation_loader:
```
Comment (nickjong) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Doesn't this iterate through the validation loader for _every_ mini-batch (training step)?

Comment (nickjong) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Doesn't the validation_loader need to be reset? How does it iterate through the data more than once?

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 6 above)
Comment (TobyRoseman) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: If consistency with other toolkits is our priority, then I don't think we should have a `seed` parameter here. Other toolkits take a `seed` parameter but I believe all of them will produce the same model given the same seed and the same dataset. That isn't true here.

Comment (nickjong) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: If fixing the seed doesn't make model creation deterministic, then I don't think we should expose it, I agree

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 9 above)
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: You're right. The validation loader in fact runs the data through only one epoch throughout training. Fixing this now.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 6 above)
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Fixed.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 8 above)
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Fixed.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 7 above)
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Fixed.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 10]
```diff
@@ -211,6 +210,27 @@ def create(input_dataset, target, feature=None, validation_set='auto',
 
     train_accuracy = _mx.metric.Accuracy()
     validation_accuracy = _mx.metric.Accuracy()
+    
+    def compute_validation_accuracy(validation_accuracy):
+        validation_loader = _SFrameClassifierIter(validation_dataset, batch_size,
```
Comment (nickjong) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: You should also be able just to create the iterator once and call `reset` on it, I think

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 11]
```diff
@@ -196,39 +208,79 @@ def update_progress(cur_loss, iteration):
...
+                z = model(x)
+                # Computes softmax cross entropy loss.
+                loss = softmax_cross_entropy(z, y)
+                # Backpropagate the error for one iteration.
+                loss.backward()
+                train_outputs.append(z)
+            # Compute training accuracy
+            train_accuracy.update(train_batch_label, train_outputs)
+
+        compute_validation_accuracy(validation_accuracy)
```
Comment (nickjong) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: I believe you're still doing this after every mini-batch

Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Fixed.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 10 above)
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Fixed.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 12]
```diff
@@ -194,39 +206,72 @@ def update_progress(cur_loss, iteration):
...
-            loss = softmax_cross_entropy(output, label)
-        loss.backward()
-        # update parameters
-        trainer.step(1)
+            for x, y in zip(train_batch_data, train_batch_label):
+                z = model(x)
+                # Computes softmax cross entropy loss.
+                loss = softmax_cross_entropy(z, y)
+                # Backpropagate the error for one iteration.
+                loss.backward()
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: I'm implementing similar functionality for the sound classifier and in my testing I noticed an interesting bug. I believe your code also has the same issue. Perhaps it's also in other toolkits as well....

Try passing in the same SFrame to `create(...)` as both the `input_dataset` and the `validation_set`. The training accuracy and the validation accuracy columns that get printed out should match, since they're the same SFrame, but I don't think they will.

I don't think it's correct to save the predicted labels while you're still doing backward passes. You probably want to do all your backward passes for in one loop and then iterate over your training set again and save the predictions.

Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Fixed. Thanks for pointing this out!
